src/cryptsetup/cryptsetup-tokens/luks2-pkcs11.c

```diff
@@ -79,9 +80,45 @@ static int luks2_pkcs11_callback(
         if (r < 0)
                 return r;
 
-        r = pkcs11_token_find_private_key(m, session, uri, &object);
+        r = pkcs11_token_find_private_key(m, session, uri, &object, &type);
         if (r < 0)
                 return r;
+#if HAVE_OPENSSL
+        if (type == CKK_EC) {
+                _cleanup_(EVP_PKEY_freep) EVP_PKEY *pkey = NULL;
+                _cleanup_(erase_and_freep) uint8_t *shared_secret = NULL;
+                size_t shared_secret_len = 0;
+                d2i_PUBKEY(&pkey, (const uint8_t **)&data->encrypted_key, data->encrypted_key_size);
+                if (!pkey)
+                        return -EIO;
+
+                r = pkcs11_token_ecdh_derive_shared_secret(m, session, object, pkey, NULL, &shared_secret_len);
+                if (r < 0)
+                        return r;
+                if (shared_secret_len == 0)
+                        return -EINVAL;
+
+                shared_secret = malloc(shared_secret_len);
+                if (!shared_secret)
+                        return log_oom();
+
+                r = pkcs11_token_ecdh_derive_shared_secret(m, session, object, pkey, shared_secret, &shared_secret_len);
+                if (r < 0)
+                        return r;
+
+                data->decrypted_key = malloc(72);
+                if (!data->decrypted_key)
+                        return log_oom();
```

Comment:
logging is not right here.

in systemd we want that functions either log about all errors, or about none (in which case the errors are propagated so that the caller can do that). usually we want more library-like utility functions to be non-logging, and main-program-like to be logging, but the line is blurry. What matters that you figure out for each function whether it better should be logging or non-logging and then stick to it. (in this context logging at LOG_DEBUG leve is not considered "real" logging and is always ok). I this function here, you handle some errors without logging, but others you do log (i.e. log_oom() here is of the latter kind, but further up you propagate -EIO/-EINVAL without logging.)

Now that is the general rule, but of course there are always exceptions and sometimes things become messy. These cryptsetup plugins for example should probably log via crypt_log_error(), i.e. libcryptsetup's logging infrastructure...

Anyway, it#s a bit messy, but please don't make this worse than it already is, and thus please stick to crypt_log_error() based logging in the plugin, and follow the other rule everywhere else.